resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,56 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        boolean directSearch = parent.searchDomains().length == 0 || StringUtil.endsWith(hostname, '.');
+        if (directSearch) {
+            internalResolve(promise);
+        } else {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
+                            int count;
+                            @Override
+                            public void operationComplete(Future<T> future) throws Exception {
+                                if (future.isSuccess()) {
+                                    original.trySuccess(future.getNow());
+                                } else {
+                                    if (count < parent.searchDomains().length) {
+                                        String searchDomain = parent.searchDomains()[count++];
+                                        Promise<T> nextPromise = parent.executor().newPromise();
+                                        String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
+                                        DnsNameResolverContext<T> nextContext = newResolverContext(parent,
+                                            nextHostname, resolveCache);
+                                        nextContext.internalResolve(nextPromise);
+                                        nextPromise.addListener(this);
+                                    } else {
+                                        original.tryFailure(future.cause());
+                                    }
+                                }
+                            }
+                        });
+                    }
+                }
+            });
+            int idx = hostname.length();
+            int dots = 0;
+            while (idx-- > 0) {
+                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
+                    internalResolve(promise);
```

Comment:
I don’t follow here - can you clarify please ?

> On Jul 7, 2016, at 11:57 PM, Scott Mitchell <notifications@github.com> wrote:
> 
> In resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java <https://github.com/netty/netty/pull/5496#discussion_r69995158>:
> 
> > +                                        nextContext.internalResolve(nextPromise);
> > +                                        nextPromise.addListener(this);
> > +                                    } else {
> > +                                        original.tryFailure(future.cause());
> > +                                    }
> > +                                }
> > +                            }
> > +                        });
> > +                    }
> > +                }
> > +            });
> > +            int idx = hostname.length();
> > +            int dots = 0;
> > +            while (idx-- > 0) {
> > +                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
> > +                    internalResolve(promise);
> can we move the promise/listener creation and adding the listener into this if block? then below just do promise.tryFailure(new UnknownHostException(hostname)) (where promise is the one in the parameter).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/pull/5496/files/50aca59178618082f484c1190a51a7c41a096c87#r69995158>, or mute the thread <https://github.com/notifications/unsubscribe/AANxirgzvIBm2fS7EWP2FdWTKSuU6ytTks5qTXZXgaJpZM4JEKKu>.
>